### DISCUSSION TITLE
Highlighter optimization for RangeOnly scans.

### DISCUSSION THREAD

**Discussion by MShark67** [TurboPack/SynEdit#188] [Ideas]
The ScanFrom and ScanRanges methods of SynEdit both scan through lines with the intent of setting each line's "Ranges" property.  If a highlighter knew that it was being asked to do a "range only" scan, it could optimize for that need. A very quick improvement with the SQL highlighter shows roughly a 50% improvement in scan time with this change (and I'm sure it could be more.)

Proposal:
Add a protected boolean field called FRangeOnly to TSynCustomHighlighter.  This would be false by default and only set in calls to SetLine with a default value of False.  e.g.:
    procedure SetLine(const Value: string; LineNumber: Integer; RangeOnly: Boolean = False); virtual;
No current highlighter overrides this method, so no highlighters would need be changed.
Then modify ScanFrom and ScanRanges to set this flag to true in their SetLine calls.
This should have no affect at all unless a highlighter has been modified to handle FRangeOnly.

Any thoughts or feedback appreciated!  If it seems a good idea, I'll pr with the changes along with an optimized SQL highlighter.  Other highlighters could be optimized on a case-by-case basis.

**Comment by pyscripter** [comment]
@MShark67 Good to hear from you.   Have you seen the two commits in relation to:
- Fonts
- Word wrap with highlighting?

Regarding your proposal I am not sure you need to change SetLine.  Both ScanRange and ScanFrom use NextToEOL.   When NextToEOL is called we are only interested in the the last token and the range info.   You could set a protected variable in NextToEOL e.g.  (FScanningToEOL) to indicate that Next is called from NextToEOL and do whatever optimizations based on that variable.

procedure TSynCustomHighlighter.NextToEol;
begin
  FScanningToEOL := True;
  while not GetEol do Next;
  FScanningToEOL := False;
end;

**Reply by MShark67** [reply]
I saw both, both great!  I've tested all those fonts and they work perfectly.  I don't actually use Wordwrap highlighted in general, so I didn't do any specific testing on that.   I didn't realize that about NextToEOL, that's even better!  Thanks!

**Reply by pyscripter** [reply]
> I don't actually use Wordwrap highlighted in general, so I didn't do any specific testing on that.

It made painting with Wordwrap and highlighter 20 - 50 times faster with very long lines and resolved a long standing todo item.

**Reply by MShark67** [reply]
>  20 - 50 times faster
Yowza!  That's awesome!
